src/main.rs

```diff
@@ -228,65 +236,110 @@ fn main() {
+            .unwrap();
         return;
     }
 
     if command == "render" {
         let angle: f64 = args[0].parse::<f64>().unwrap();
         let nwidth: usize = args[1].parse::<usize>().unwrap();
         let nodepressions: bool = args.len() > 2 && args[2] == "nodepressions";
-        pullauta::render::render(&config, &thread, &tmpfolder, angle, nwidth, nodepressions)
-            .unwrap();
+        pullauta::render::render(
+            &fs,
+            &config,
+            &thread,
+            &tmpfolder,
+            angle,
+            nwidth,
+            nodepressions,
+        )
+        .unwrap();
         return;
     }
 
     let proc = config.processes;
     if command.is_empty() && batch && proc > 1 {
-        let mut handles: Vec<thread::JoinHandle<()>> = Vec::with_capacity((proc + 1) as usize);
-        for i in 0..proc {
-            let config = config.clone();
-            let handle = thread::spawn(move || {
-                info!("Starting thread");
-                pullauta::process::batch_process(&config, &format!("{}", i + 1));
-                info!("Thread complete");
-            });
-            thread::sleep(time::Duration::from_millis(100));
-            handles.push(handle);
-        }
-        for handle in handles {
-            handle.join().unwrap();
+        if config.experimental_use_in_memory_fs {
+            // copy all the input files into the memory file system
+            let fs = pullauta::io::fs::memory::MemoryFileSystem::new();
+            fs.create_dir_all(&config.lazfolder).unwrap();
+            for file in fs::read_dir(&config.lazfolder).unwrap() {
+                let file = file.unwrap();
+                let path = file.path();
+                println!("Copying {} into memory fs", path.display());
+                fs.load_from_disk(&path, &path).unwrap();
+            }
+
+            // do the processing
+            let mut handles: Vec<thread::JoinHandle<()>> = Vec::with_capacity((proc + 1) as usize);
+            for i in 0..proc {
+                let config = config.clone();
+                let fs = fs.clone();
+                let handle = thread::spawn(move || {
+                    info!("Starting thread");
+                    pullauta::process::batch_process(&config, &fs, &format!("{}", i + 1));
+                    info!("Thread complete");
+                });
+                thread::sleep(time::Duration::from_millis(100));
+                handles.push(handle);
+            }
+            for handle in handles {
+                handle.join().unwrap();
+            }
+
+            // copy the output files back to disk
+            for path in fs.list(&config.batchoutfolder).unwrap() {
+                info!("Copying {} from memory fs to disk", path.display());
+                fs.save_to_disk(&path, &path).unwrap();
+            }
+        } else {
+            let mut handles: Vec<thread::JoinHandle<()>> = Vec::with_capacity((proc + 1) as usize);
```

Comment:
Yes it's the same code but using different types, hence the duplication (otherwise there would be compiler errors for mismatching types). I'll see what I can do though to clean it up 👍

Reply:
Reduced some code duplication in the latest commit :100:

Reply:
If you mean something like this;
```rust
       if config.experimental_use_in_memory_fs {
            // copy all the input files into the memory file system
            let fs = pullauta::io::fs::memory::MemoryFileSystem::new();
            fs.create_dir_all(&config.lazfolder).unwrap();
            for file in fs::read_dir(&config.lazfolder).unwrap() {
                let file = file.unwrap();
                let path = file.path();
                println!("Copying {} into memory fs", path.display());
                fs.load_from_disk(&path, &path).unwrap();
            }
        }

        // do the processing
        let mut handles: Vec<thread::JoinHandle<()>> = Vec::with_capacity((proc + 1) as usize);
        for i in 0..proc {
            let config = config.clone();
            let fs = fs.clone();
            let handle = thread::spawn(move || {
                info!("Starting thread");
                pullauta::process::batch_process(&config, &fs, &format!("{}", i + 1));
                info!("Thread complete");
            });
            thread::sleep(time::Duration::from_millis(100));
            handles.push(handle);
        }
        for handle in handles {
            handle.join().unwrap();
        }

        if config.experimental_use_in_memory_fs {
            // copy the output files back to disk
            std::fs::create_dir_all(&config.batchoutfolder).unwrap();
            for path in fs.list(&config.batchoutfolder).unwrap() {
                info!("Copying {} from memory fs to disk", path.display());
                fs.save_to_disk(&path, &path).unwrap();
            }
        }
```
I agree it would be nice, but there are some problems. First the initialization of the in-memory storage is lost inside the closure, and if we try to mark `fs` as mutable we get a type error since the Local and InMemory types are different. We would have to do some `Box<dyn FileSystem>` but  the `FileSystem` trait is not object safe (it has functions that return `impl Read`, for example) so it's not even possible :disappointed: So this is the best I could come up with :laughing: If I misunderstood please provide some example to make your point clearer :100: